### PR TITLE
Fix use quota bug

### LINKED ISSUE #31174: Quota limit was not reached, but the limit was exceeded
## Describe what's wrong
quota limit was not reached, but the limit was exceeded，Changes from https://github.com/ClickHouse/ClickHouse/commit/d8d2bd885c72ae06707f0a15001f2bfb7ba21054#diff-ebadfab5af066c0ff22a9999d1455c293936dff46dd572f002b912ab355bf00f

## Does it reproduce on recent release?
Version 21.3 or later

## How to reproduce
1、Set the quota of the account
```
CREATE USER IF NOT EXISTS jd_olap IDENTIFIED WITH double_sha1_password BY '123456'

GRANT SHOW, SELECT, INSERT, ALTER, CREATE DATABASE, CREATE TABLE, CREATE VIEW, CREATE DICTIONARY, DROP, TRUNCATE, OPTIMIZE, SYSTEM MERGES, SYSTEM TTL MERGES, SYSTEM FETCHES, SYSTEM MOVES, SYSTEM SENDS, SYSTEM REPLICATION QUEUES, SYSTEM SYNC REPLICA, SYSTEM RESTART REPLICA, SYSTEM FLUSH DISTRIBUTED, dictGet ON jd_olap.* TO jd_olap

CREATE QUOTA jd_olap_10s FOR INTERVAL 10 second MAX queries = 10 TO jd_olap;
```

2、create table
```
CREATE DATABASE IF NOT EXISTS quota_test_local on cluster system_cluster;

CREATE TABLE IF NOT EXISTS jd_olap.quota_test_local on cluster system_cluster 
( 
`user` String, 
`max_concurrent_queries` UInt32, 
`max_execution_time` UInt32, 
`requests_per_minute` UInt32,
`dt` Date 
) 
ENGINE = ReplicatedReplacingMergeTree('/clickhouse/system_cluster/jdob_ha/jd_olap/quota_test_local/{shard}', '{replica}') \
PARTITION BY dt 
ORDER BY user 
SETTINGS storage_policy = 'jdob_ha', index_granularity = 8192;
```

3、execute the query（insert， select， alter）, will trigger the problem
image

## Expected behavior**
there is a bug where end time is 1970, causing end_loaded.count =0；
image

the used quota is not reset when it is first calculated (e.g., when a node is restarted or created for the first time) because 1970 is the first end time, the initial used logic is incremented from the first calculated value to the next interval. Used quota values are reset only at the next interval, as must be shown in the following figure: 16:45 and 16:47 are both calculated for the first time on the node restart, and the quota is not reset；
```
2021.11.09 16:49:17.481461 [ 14825 ] {1686DC283041A762} <Information> EnabledQuota: current_time: 2021-11-09 16:49:17, end: 1970-01-01 08:00:00
2021.11.09 16:49:17.481479 [ 14825 ] {1686DC283041A762} <Information> EnabledQuota: need_reset_counters: false, end_load_count: 0, interval.used: 11
```

### PR DESCRIPTION
**Pull request by sunny19930321**
Changelog category (leave one):

Bug Fix

Changelog entry:
Quota limit was not reached, but the limit was exceeded. This PR fixes [#31174](https://github.com/ClickHouse/ClickHouse/issues/31174)

this PR https://github.com/ClickHouse/ClickHouse/pull/31337
this change does not solve the problem，need_reset_counters the first time is still false，I think the first time the limit is exceeded, used should be reset to 0 @vitlibar

### CODE REVIEW

File: src/Access/EnabledQuota.cpp [hunk 1]
```diff
@@ -67,7 +67,7 @@ struct EnabledQuota::Impl
             {
                 /// We reset counters only if the interval's end has been calculated before.
                 /// If it hasn't we just calculate the interval's end for the first time and don't reset counters yet.
-                need_reset_counters = (end_loaded.count() != 0);
+                need_reset_counters = (end_of_interval.load().count() != 0);
```
Comment (vitlibar) [inline src/Access/EnabledQuota.cpp:70]: I finally understand the problem you want to fix.

Let's change this line to a simplier equivalent

```
need_reset_counters = true;
```

and in order to fix the tests in more correct way please add the following lines:

```
if (!interval.end_of_interval.load().count())
{
    /// We need to calculate end of the interval if it hasn't been calculated before.
    bool dummy;
    getEndOfInterval(interval, current_time, dummy);
}
```

before [this](https://github.com/ClickHouse/ClickHouse/blob/9892151290bee34f068be82355a8c0b54e67fb88/src/Access/EnabledQuota.cpp#L98) line and also before [this](https://github.com/ClickHouse/ClickHouse/blob/9892151290bee34f068be82355a8c0b54e67fb88/src/Access/EnabledQuota.cpp#L126).

File: src/Core/Settings.h [hunk 2]
```diff
@@ -641,10 +641,17 @@ class IColumn;
 // End of FORMAT_FACTORY_SETTINGS
 // Please add settings non-related to formats into the COMMON_SETTINGS above.
 
+// settings work in query runtime
+#define RUNTIME_QUERY_SETTINGS(M) \
+    M(Bool, is_reinterpreted_execution, false, "Queries such as show tables will be reinterpreted to select query.", 0)
+// End of RUNTIME_QUERY_SETTINGS
```
Comment (vitlibar) [inline src/Core/Settings.h:647]: This new setting looks too complex for this easy task, the previous solution with

```
bool ignoreQuota() const override { return false; }
```

looked more clear and obvious to me.

Comment (sunny19930321) [reply src/Core/Settings.h:647]: ok

File: src/Access/EnabledQuota.cpp (hunk 1 above)
Comment (sunny19930321) [reply src/Access/EnabledQuota.cpp]: ok

File: tests/integration/test_quota/test.py [hunk 3]
```diff
@@ -393,30 +393,45 @@ def test_query_inserts():
 
 
 def test_consumption_of_show_tables():
+    assert re.match(
+        "myQuota\\tdefault\\t.*\\t31556952\\t0\\t1000\\t0\\t.*\\t\\\\N.*",
+        instance.query("SHOW QUOTA"))
```
[the remark this answers is not in the document]
Comment (vitlibar) [reply tests/integration/test_quota/test.py:398]: Let me help you

**Comment by vitlibar** [comment]
@Mergifyio update

File: tests/integration/test_quota/test.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (vitlibar) [reply tests/integration/test_quota/test.py]: This PR along with https://github.com/ClickHouse/ClickHouse/pull/32575 must fix the issue.

[the remark this answers is not in the document]
Comment (sunny19930321) [reply tests/integration/test_quota/test.py]: tks，the problem has been solved

**Pull request merged**